Fix add/del user to group modifies LDAP group

piernov opened the pull request:
Bug present at least in 12.0, 13.0 and develop but it has probably been there for longer.

_**v2:**_
**PR based off the 12.0 branch but will have to be merged in newer branches as well.**

# Fix add/del user to group modifies LDAP group 

Adding or removing a user from a group modifies the user object on Dolibarr's side.
In LDAP however, members of a group are stored in the group itself.
Therefore group must be updated after adding/removing a user from it.
Update group in LDAP with new list of users at the end of USER_MODIFY trigger.

~~_**v1:**_~~
~~**PR based off the 12.0 branch (to at least have the correct `USERGROUP_MODIFY` trigger) but will have to be merged in newer branches as well.**~~

# ~~Fix add/del user to group trigger USERGROUP_MODIFY~~

~~Adding or removing a user from a group modifies the user object on Dolibarr's side.~~
~~In LDAP however, members of a group are stored in the group itself.~~
~~Therefore group must be updated after adding/removing a user from it.~~
~~Trigger the USERGROUP_MODIFY event from SetInGroup/RemoveFromGroup to that effect.~~

~~**Warning**: I don't know if there are side effects to calling the trigger here. I also don't know if it is the proper way to solve this problem.~~

atm-maxime commented:
Hi ! Thanks for contributing in Dolibarr and welcome !
For this one let's see what @eldy says, but maybe new triggers like USER_ADD_IN_GROUP / USER_REMOVE_FROM_GROUP would be better here as that's neither a user modification, nor a group modification.

eldy commented:
In very old versions of dolibarr (< v6), we used a trigger called USER_SETINGROUP andd USER_REMOVEFROMGROUP (code is still present and commented into htdocs/core/triggers/interface_50_modLdap_Ldapsynchro.class.php).
However, they were removed to better follow the rule: 1 Trigger for 1 CRUD events (so Create, Read, Update or Delete).

So for this need, you must keep only one trigger call, the call of USER_MODIFY (no trigger on group must be done because nothing is changed on group, we change something on links between user and groups, and choice for that was to include this into the event of modification of USER).

What you must do is modifying the code inside the trigger USER_MODIFY into the trigger of Ldap module, so into the file htdocs/core/triggers/interface_50_modLdap_Ldapsynchro.class.php.

Just add a test into USER_MODIFY part of the LDAP trigger:
```
if !empty($this->context['newgroupid']) {      // We are in context of adding a new group to user
  // call the same code that is inside the USERGROUP_MODIFY part of trigger.
}
if !empty($this->context['oldgroupid']) {      // We are in context of remobing a group from user
  // call the same code that is inside the USERGROUP_MODIFY part of trigger.
}
```

Can you fix that way ?

piernov committed:
Fix add/del user to group modifies LDAP group

Adding or removing a user from a group modifies the user object on Dolibarr's side.
In LDAP however, members of a group are stored in the group itself.
Therefore group must be updated after adding/removing a user from it.
Update group in LDAP with new list of users at the end of USER_MODIFY trigger.

piernov commented:
Thanks, that makes sense.
I updated the PR by removing the previous changes and instead adapting the `USER_MODIFY` trigger with the code from the `USER_SETINGROUP` trigger, for both cases of adding and removing a group from a user.
The issue is that it creates a bunch of duplicate lines of code.
Each trigger handling already has some similar duplicate code so I am not sure if it is ok or not.
